### ISSUE TITLE
[tt-train] vocab_parallel_cross_entropy_loss holds ~788 MB of dead full-vocab intermediates in backward

### ISSUE THREAD

**Issue by imichalakTT** [tenstorrent/tt-metal#52923]
`ttml::ops::distributed::vocab_parallel_cross_entropy_loss` (`tt-train/sources/ttml/ops/distributed/losses.cpp`)
allocates several full-vocab intermediates and frees none early.

Sizes below are llama8b_tp4 (B=1, S=2048, local_V=32064), where a full-vocab
`[B,1,S,local_V]` tensor is 131 MB in bf16 and 263 MB in fp32.

### Memory

The whole backward is this chain:

```cpp
auto local_exp = fused_subtract_exp_fp32(logits->get_value(), global_max);        // fp32, 263 MB
auto softmax_k = ttnn::multiply(local_exp, ttnn::reciprocal(global_sum));         // fp32, 263 MB
auto scaled_softmax = ttnn::multiply(softmax_k, inv_N, ttnn::DataType::BFLOAT16);  // bf16, 131 MB

ttnn::prim::ttml_subtract_at_target(scaled_softmax, targets_raw, local_V, cluster_axis,
                                    /*first_v=*/0U, scaled_softmax, /*subtract_value=*/inv_N);

logits->add_grad(ttnn::multiply(scaled_softmax, out->get_grad(), ttnn::DataType::BFLOAT16));  // bf16, 131 MB
```

All four are C++ locals in the grad closure, so all four are still live when the last
`multiply` allocates: ~788 MB. `local_exp` is dead once `softmax_k` exists and `softmax_k`
is dead once `scaled_softmax` exists, but neither is released. The working set the last
line actually needs is closer to the retained logits (131 MB) plus one output (131 MB).

The loss closure is the first to run in backward, so no forward activation has been
released yet and these four tensors stack on the full activation peak. They sit on the
global memory peak, which sets max trainable sequence length and what fits at a given
parallelism — removing one measurably shifts that limit.

`softmax_k` exists only to carry a scaling that is then applied a second time. Both factors
— `reciprocal(global_sum)` and `inv_N` — are `[B,1,S,1]` or scalar, so a full-vocab FP32
tensor is materialized to hold the intermediate of a two-step scale.

Forward has the mirror problem. It computes the same `local_exp`, consumes it immediately
in one reduction, and then never reads it again:

```cpp
auto local_exp = fused_subtract_exp_fp32(logits->get_value(), global_max);        // fp32, 263 MB
auto local_sum = ttnn::sum(local_exp, 3, /*keepdim=*/true, std::nullopt,
                           core::ComputeKernelConfig::precise());
auto global_sum = ttnn_fixed::distributed::all_reduce(local_sum, cluster_axis);
// local_exp is never read again, but lives until the function returns
```

It stays resident through two collectives, a layout conversion, `select_target_logit` and
several reductions — and backward then pays a full recompute to rebuild it. It is neither
saved nor promptly freed.

Both paths materialize the full `[B,1,S,local_V]` in FP32, while the precision requirement
is on the reduction (the `ttnn::sum` above requests `ComputeKernelConfig::precise()`).

### Device time

Secondary and needs profiling: roughly nine full-vocab passes per step, plus three
separate collectives in forward (an all-gather of the local maxes and two all-reduces)
on payloads of about 131 KB each.
